### PR TITLE
Layer extension optimization

### PR DESCRIPTION
**Pull request by StephenHodgson**
Overview
---
There's lots of unnecessary memory allocations when sorting layers.
Changed the `IEnumerable` parameters to a basic array.
Updated Focus Manager assertions to remove memory allocations when getting the type even if no assertion is thrown.
Removed the list of `ComparableRaycastResult` in `EventSystemExtensions`, because we don't need it.

### CODE REVIEW

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs [hunk 1]
```diff
@@ -552,8 +552,8 @@ private void RaycastPhysics(PointerData pointer, LayerMask[] prioritizedLayerMas
             RayStep rayStep = default(RayStep);
             RaycastHit physicsHit = default(RaycastHit);
 
-            Debug.Assert(pointer.PointingSource.Rays != null, "No valid rays for " + pointer.GetType());
```
Comment (keveleigh) [inline Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: Doesn't this remove context from the assertion statement? Why would we want to remove that?

File: Assets/HoloToolkit/Common/Scripts/Extensions/LayerExtensions.cs [hunk 2]
```diff
@@ -54,15 +53,16 @@ public static int Activation
         /// <param name="layer">Layer to search for</param>
         /// <param name="layerMasks">List of LayerMasks to search</param>
         /// <returns>LayerMaskList index, or -1 for not found</returns>
-        public static int FindLayerListIndex(this int layer, IEnumerable<LayerMask> layerMasks)
+        public static int FindLayerListIndex(this int layer, LayerMask[] layerMasks)
```
Comment (keveleigh) [inline Assets/HoloToolkit/Common/Scripts/Extensions/LayerExtensions.cs]: Changing the parameters of a method would be a breaking change, right? Is this change worth that?

Comment (StephenHodgson) [reply Assets/HoloToolkit/Common/Scripts/Extensions/LayerExtensions.cs]: No, it's not a breaking change.

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs (hunk 1 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: `GetType()` is causing an enormous amount of allocations during this time.

File: Assets/HoloToolkit/Common/Scripts/Extensions/LayerExtensions.cs (hunk 2 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit/Common/Scripts/Extensions/LayerExtensions.cs]: Yes, this change is very worth it, considering it removes _all_ allocations for this check.

**Comment by StephenHodgson** [comment]
@NeerajW  It does:
image

Sorry to have dismissed your review.

**Comment by StephenHodgson** [comment]
@Zod- any chance I could get your eyes on this as well?

File: Assets/HoloToolkit/Common/Scripts/Extensions/EventSystemExtensions.cs [hunk 3]
```diff
@@ -25,17 +24,26 @@ public static RaycastResult Raycast(this EventSystem eventSystem, PointerEventDa
...
-        private static RaycastResult PrioritizeRaycastResult(LayerMask[] layerMaskPrio)
+        private static RaycastResult PrioritizeRaycastResult(LayerMask[] priority)
         {
-            ComparableRaycastResults.Clear();
-            foreach (var raycastResult in RaycastResults)
+            ComparableRaycastResult maxResult = default(ComparableRaycastResult);
+
+            for (var i = 0; i < RaycastResults.Count; i++)
             {
-                if (raycastResult.gameObject == null) { continue; }
```
Comment (Zod-) [inline Assets/HoloToolkit/Common/Scripts/Extensions/EventSystemExtensions.cs]: I'd keep `var raycastResult = RaycastResults[i];`

Comment (StephenHodgson) [reply Assets/HoloToolkit/Common/Scripts/Extensions/EventSystemExtensions.cs]: Good catch

**Comment by StephenHodgson** [comment]
@Railboy had to update the test scene on @Zod-'s request.  Should be good now.

**Comment by Zod-** [comment]
Do we need the clear here? Atleast when running the tests it doesn't seem to care if we do it or not.
```cs
RaycastResults.Clear();
eventSystem.RaycastAll(pointerEventData, RaycastResults);
return PrioritizeRaycastResult(layerMasks);
```

**Comment by Zod-** [comment]
Ok, I've checked the source of `EventSystem.RaycastAll`. The first thing it does is clear the results itself so we don't have to do that.

https://bitbucket.org/Unity-Technologies/ui/src/6a975b5d7ec71f5cf99b615bc31d1d14e6e246db/UnityEngine.UI/EventSystem/EventSystem.cs?at=2017.1&fileviewer=file-view-default#EventSystem.cs-182

File: Assets/HoloToolkit/Common/Scripts/Extensions/LayerExtensions.cs (hunk 2 above)
Comment (keveleigh) [reply Assets/HoloToolkit/Common/Scripts/Extensions/LayerExtensions.cs]: It's a public method though, so devs might call it in their apps too.

What/how much do we gain from removing all allocations?

File: Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/HoloToolkit/Input/Scripts/Focus/FocusManager.cs]: Do you have a measured performance increase?

**Comment by StephenHodgson** [comment]
Where do we call clear?

**Comment by StephenHodgson** [comment]
@keveleigh, sorry made one last tweak per @Zod-

**Comment by keveleigh** [comment]
@StephenHodgson Sounds good to me :)

**Comment by StephenHodgson** [comment]
Should be good now.

**Pull request merged**